File: messagearea.rb
require 'gtk2'

module Cpg
	class MessageArea < Gtk::HBox
		type_register
		
		attr_reader :action_area
		
		def initialize
			super({'homogeneous' => false, 'spacing' => 0})
			
			self.app_paintable = true
			
			@mainbox = Gtk::HBox.new(false, 16)
			@mainbox.border_width = 8

			@contents = nil
			@action_area = Gtk::VBox.new(TRUE, 10)
			
			@mainbox.pack_end(@action_area, false, true, 0)
			self.pack_start(@mainbox, true, true, 0)

			@changing_style = false
			
			@mainbox.signal_connect('style_set') { |w, prevstyle| do_style_set(prevstyle) }
			signal_connect('expose_event') { |w, ev| paint_message_area(ev) }
		end
		
		def contents=(c)
			@contents.destroy if @contents
			@contents = c
			
			@mainbox.pack_start(c, true, true, 0)
			c.show
		end
		
		def actions=(actions)
			@action_area.children.each { |x| x.destroy }
			
			actions.each do |action|
				@action_area.pack_start(action[0], false, false, 0)
				action[0].show

				if action[0].is_a?(Gtk::Button)
					action[0].signal_connect('clicked') { |b| action[1].call }
				end
			end
		end
		
		def do_style_set(prevstyle)
			return if @changing_style

			w = Gtk::Window.new(Gtk::Window::POPUP);
			w.name = "gtk-tooltip"
			
			w.ensure_style

			@changing_style = true
			self.set_style(w.style)
			@changing_style = false
			
			w.destroy
			queue_draw
		end
		
		def paint_message_area(event)
			self.style.paint_flat_box(self.window,
									  Gtk::STATE_NORMAL,
									  Gtk::SHADOW_OUT,
									  nil,
									  self,
									  "tooltip",
									  self.allocation.x + 1,
									  self.allocation.y + 1,
									  self.allocation.width - 2,
									  self.allocation.height - 2)
			false
		end
	end
end

File: Cdn.Studio.Widgets/MessageArea.cs
using System;
using Gtk;

namespace Cpg.Studio.Widgets
{
	public class MessageArea : HBox
	{
		private HBox d_mainBox;
		private VBox d_actionArea;
		private bool d_changingStyle;
		private Widget d_contents;
		private Label d_secondary;
		
		public delegate void ResponseHandler(object source, ResponseType type);
		public event ResponseHandler Response = delegate {};
		
		public MessageArea(params object[] actions)
		{
			d_mainBox = new HBox(false, 16);
			d_mainBox.BorderWidth = 8;
			d_mainBox.Show();
			
			d_actionArea = new VBox(false, 10);
			d_actionArea.Show();
			
			d_mainBox.PackEnd(d_actionArea, false, true, 0);
			
			PackStart(d_mainBox, true, true, 0);
			
			AppPaintable = true;
			CanFocus = true;
			
			d_mainBox.StyleSet += OnStyleSet;
			
			for (int i = 0; i < actions.Length; ++i)
			{
				if (i + 1 < actions.Length)
				{
					AddAction(actions[i] as string, (ResponseType)actions[i + 1]);
				}
			}
		}
		
		public static MessageArea Create(string icon, string primary, string secondary, params object[] actions)
		{
			MessageArea ret = new MessageArea(actions);
			
			Widget content;
			VBox vbox = new VBox(false, 6);
			
			if (icon != null)
			{
				HBox hbox = new HBox(false, 12);
				hbox.PackStart(new Image(icon, IconSize.Dialog), false, false, 0);
				hbox.PackStart(vbox, true, true, 0);
				
				content = hbox;
			}
			else
			{
				content = vbox;
			}	
			
			Label prim = new Label("<b>" + System.Security.SecurityElement.Escape(primary) + "</b>");
			prim.UseMarkup = true;
			prim.Xalign = 0;
			prim.UseUnderline = false;
			
			vbox.PackStart(prim, false, true, 0);
			
			Label sec = new Label(secondary);
			sec.Xalign = 0;
			sec.UseUnderline = false;
			sec.Wrap = true;
			
			ret.d_secondary = sec;
			
			vbox.PackStart(sec, false, true, 0);
			
			content.ShowAll();
			
			ret.Contents = content;
			
			return ret;
		}
		
		public static MessageArea Create(string primary, string secondary, params object[] actions)
		{
			return Create(null, primary, secondary, actions);
		}

		public Button AddAction(string stock, ResponseType resp)
		{
			Button button = new Button(stock);
			button.CanDefault = true;
			button.Show();
					
			if (resp == ResponseType.Help)
			{
				d_actionArea.PackEnd(button, false, false, 0);
			}
			else
			{
				d_actionArea.PackStart(button, false, false, 0);
			}
			
			button.Data["MessageAreaResponse"] = resp;
			
			button.Clicked += delegate (object source, EventArgs args)
			{
				EmitResponse(resp);
			};
			
			return button;
		}
		
		public void SetDefaultResponse(ResponseType resp)
		{
			foreach (Widget w in d_actionArea.Children)
			{
				if (!w.Data.ContainsKey("MessageAreaResponse"))
					continue;
					
				if ((ResponseType)w.Data["MessageAreaResponse"] == resp)
				{
					w.GrabDefault();
				}
			}
		}
		
		public void SetResponseSensitive(ResponseType resp, bool sensitive)
		{
			foreach (Widget w in d_actionArea.Children)
			{
				if (!w.Data.ContainsKey("MessageAreaResponse"))
					continue;
					
				if ((ResponseType)w.Data["MessageAreaResponse"] == resp)
				{
					w.Sensitive = false;
				}
			}
		}
		
		public VBox ActionArea
		{
			get
			{
				return d_actionArea;
			}
		}
		
		public Widget Contents
		{
			get
			{
				return d_contents;
			}
			set
			{
				if (d_contents != null)
				{
					d_mainBox.Remove(d_contents);
				}
				
				d_contents = value;
				d_mainBox.PackStart(d_contents, true, true, 0);
			}
		}

		void OnStyleSet(object o, StyleSetArgs args)
		{
			if (d_changingStyle)
			{
				return;
			}
			
			Gtk.Window wnd = new Gtk.Window(WindowType.Popup);
			wnd.Name = "gtk-tooltip";
			wnd.EnsureStyle();
			
			Gtk.Widget thisone = (Gtk.Widget)o;
			
			Style style = wnd.Style;
			d_changingStyle = true;
			thisone.Style = style;
			d_changingStyle = false;
			
			QueueDraw();
		}
		
		protected override void OnSizeAllocated(Gdk.Rectangle allocation)
		{
			base.OnSizeAllocated(allocation);

			if (d_secondary != null)
			{
				d_secondary.SetSizeRequest((int)(allocation.Width * 0.6), -1);
			}
		}

		protected override bool OnExposeEvent(Gdk.EventExpose evnt)
		{
			Gtk.Style.PaintFlatBox(Style,
			                       evnt.Window,
			                       StateType.Normal,
			                       ShadowType.Out,
			                       evnt.Area,
			                       this,
			                       "tooltip",
			                       Allocation.X + 1,
			                       Allocation.Y + 1,
			                       Allocation.Width - 2,
			                       Allocation.Height - 2);
			
			return base.OnExposeEvent(evnt);
		}
		
		private void EmitResponse(ResponseType resp)
		{
			Response(this, resp);
		}
		
		protected override bool OnKeyPressEvent(Gdk.EventKey evnt)
		{
			if (evnt.Key == Gdk.Key.Escape)
			{
				EmitResponse(ResponseType.DeleteEvent);
				return true;
			}
			else
			{			
				return base.OnKeyPressEvent (evnt);
			}
		}

	}
}
